### PR TITLE
Running EVE from a single binary.

### PR DESCRIPTION
**Pull request by adarsh-zededa**
Instead of spawning each EVE services with its own binary, we'll be starting EVE using a single binary zedbox.

TODO:
Fix log source.

Open question:
Do we need `/run/<service>.pid` file? Since all the service will have the same PIDs.

Signed-off-by: adarsh-zededa <adarsh@zededa.com>

### CODE REVIEW

**Comment by deitch** [comment]
@adarsh-zededa I don't get this. Are we saying that each service - e.g. volumemgr, verifier, downloader, zedagent, etc. etc. - are now just functions (or maybe goroutines) in a single running process?

Put in other terms, did pillar just become a monolith? I mean it not in a derogatory sense - there are plenty of cases where monoliths make sense - but as a description of an architectural pattern.

Most specifically:

* do we really need the complex pubsub if that is what we are doing, since we no longer have any form of IPC, but rather IFC (inter-function communication), which all can happen in-memory?
* more importantly (by far), does this not break the security model? We always were careful that, e.g. one process has access to the Internet (zedclient and downloader, no others), verifier is standalone, etc. I think @eriknordmark has thoughts on this.

**Comment by eriknordmark** [comment]
> @adarsh-zededa I don't get this. Are we saying that each service - e.g. volumemgr, verifier, downloader, zedagent, etc. etc. - are now just functions (or maybe goroutines) in a single running process?

The mapping between pillar microservices and runtimes becomes more flexible. Prior to this each microservice had it own golang runtime, and with this they can share a single runtime. But we can also map to two runtimes (e.g., those microservices which need TCP connections to a controller can be handled by one golang runtime, while others use a runtime which is isolated from Internet.)
> Put in other terms, did pillar just become a monolith? I mean it not in a derogatory sense - there are plenty of cases where monoliths make sense - but as a description of an architectural pattern.

Not any more than it being a monolith due to using the same instance of the Linux kernel.

> Most specifically:
> 
> * do we really need the complex pubsub if that is what we are doing, since we no longer have any form of IPC, but rather IFC (inter-function communication), which all can happen in-memory?

That would remove flexibility. And what would the benefits be? Saving a millisecond or ten in the deployment time for an app instance, which takes minutes to download and deploy?

> * more importantly (by far), does this not break the security model? We always were careful that, e.g. one process has access to the Internet (zedclient and downloader, no others), verifier is standalone, etc. I think @eriknordmark has thoughts on this.

See above @deitch

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go
Comment (eriknordmark) [inline pkg/pillar/cmd/tpmmgr/tpmmgr.go:74]: I think we should generalize the done pattern to have the zedbox code touch agentname.done when the goroutine returns.

File: pkg/pillar/cmd/logmanager/logmanager.go
Comment (eriknordmark) [inline pkg/pillar/cmd/logmanager/logmanager.go:301]: This is useful, but I think it would make sense to split it off as a separate PR.
(I need this to complete the source in the logs and might refactor it a bit which would be a local change in zedcloud package.)

File: pkg/pillar/cmd/hardwaremodel/hardwaremodel.go
Comment (eriknordmark) [inline pkg/pillar/cmd/hardwaremodel/hardwaremodel.go:29]: Can we do this as a small separate PR?

File: pkg/pillar/scripts/device-steps.sh
Comment (eriknordmark) [inline pkg/pillar/scripts/device-steps.sh:10]: Generalize to /run/<agentName>.done

File: pkg/pillar/scripts/device-steps.sh
Comment (eriknordmark) [inline pkg/pillar/scripts/device-steps.sh:65]: wait_for_done $agentname $time?

**Comment by deitch** [comment]
> Not any more than it being a monolith due to using the same instance of the Linux kernel.

Well, that isn't a totally fair comparison, since multiple processes that communicate over a socket (network, whatever) can be deployed on one kernel or multiple. But that doesn't really apply to the core of the discussion here. :-)

> The mapping between pillar microservices and runtimes becomes more flexible. Prior to this each microservice had it own golang runtime, and with this they can share a single runtime. But we can also map to two runtimes (e.g., those microservices which need TCP connections to a controller can be handled by one golang runtime, while others use a runtime which is isolated from Internet.)

OK, so a few questions:

* how do we determine which are grouped together? Is it all in `device-steps.sh`? We need a very clear grouping and separating. When every single one was separate, it didn't matter, since `device-steps.sh` just started all of them as independent ones. 
* we need to document these changes before it can be merged in, and the very specific "these run together"

As far as I can see, this PR merges them all together?

**Comment by eriknordmark** [comment]
> OK, so a few questions:
> 
> * how do we determine which are grouped together? Is it all in `device-steps.sh`? We need a very clear grouping and separating. When every single one was separate, it didn't matter, since `device-steps.sh` just started all of them as independent ones.

I don't know what that policy would look like (set by the controller vs. determined based on available memory?)
But I think a policy knob/file which zedbox reads to determine 1) how many zedboxes to run and with which privillges/devices 2) when asked to start <service> knows which instance of zedbox implements that service.
That part is all TBD.

> * we need to document these changes before it can be merged in, and the very specific "these run together"

Good point.

> 
> As far as I can see, this PR merges them all together?

Correct. Which isn't bad as long as we maintain the flexibility to run them separately.

**Comment by deitch** [comment]
> I don't know what that policy would look like (set by the controller vs. determined based on available memory?)
But I think a policy knob/file which zedbox reads to determine 1) how many zedboxes to run and with which privillges/devices 2) when asked to start knows which instance of zedbox implements that service.
That part is all TBD.

I like where you're going, but I was thinking much simpler to start. Just set them arbitrarily (well, based on our knowledge) in `device-steps.sh`. We can add runtime config control at a later stage. The way this is set up now, we have eliminated completely the separation we had before.

Given the amount on our plates, and the tendency of temporary things to last forever, I suspect it will be months, if not years, before we re-separate them.

Would it not be better for this PR at least to start by having `device-steps.sh` use some grouping that captures our desired and current security model? We can add config control later, but it appears we are going from "everything separate" to "nothing separate", when we know already there are at least some things we want separate.

**Comment by eriknordmark** [comment]
> TODO:
> Fix log source.

Underway in https://github.com/lf-edge/eve/pull/1295

> Open question:
> Do we need `/run/<service>.pid` file? Since all the service will have the same PIDs.

Let's defer that until we have some experience with forcing golang panics and log.Fatals and hangs to see what the debugging experience looks like. What we have now with N pid files does detect the crashes.

Note that we should plan for more than one zedbox to be able to have some defense in depth by restricting privileges, hence it is more than just a single zedbox.pid.

File: pkg/pillar/cmd/logmanager/logmanager.go
Comment (adarsh-zededa) [reply pkg/pillar/cmd/logmanager/logmanager.go:301]: https://github.com/lf-edge/eve/pull/1318

File: pkg/pillar/cmd/hardwaremodel/hardwaremodel.go
Comment (adarsh-zededa) [reply pkg/pillar/cmd/hardwaremodel/hardwaremodel.go:29]: https://github.com/lf-edge/eve/pull/1319

File: pkg/pillar/scripts/device-steps.sh
Comment (rvs) [reply pkg/pillar/scripts/device-steps.sh:10]: +1

File: pkg/pillar/zedbox/zedbox.go
Comment (rvs) [inline pkg/pillar/zedbox/zedbox.go:86]: Again, can we retain (via non-default cmd flag) a capability of having old stype zebox behavior?

File: pkg/pillar/scripts/device-steps.sh
Comment (rvs) [reply pkg/pillar/scripts/device-steps.sh:65]: See my question above, but can't this be done by, effectively, introducing a command line utility to talk on our pubsub?

File: pkg/pillar/zedbox/zedbox.go
Comment (rvs) [reply pkg/pillar/zedbox/zedbox.go:86]: Or maybe even -- lets call new binary pillar and the old school binary will still be called zedbox. That way, you can have your cake and eat it too (build zedbox as it used to be and have pillar -> zedbox)

File: pkg/pillar/scripts/device-steps.sh
Comment (rvs) [inline pkg/pillar/scripts/device-steps.sh:154]: Another option here (if we can't wrap it into proper pubsub) would be to wait on the socket to reply with "done". That way -- one shot services like vaultmgr, etc. will reply "done" on the socket when they are done and everything else will reply immeditately.

start-service.sh will then simply block on reading from the socket and all the complexity will be gone.

Comment (eriknordmark) [reply pkg/pillar/scripts/device-steps.sh:154]: > Another option here (if we can't wrap it into proper pubsub) would be to wait on the socket to reply with "done". That way -- one shot services like vaultmgr, etc. will reply "done" on the socket when they are done and everything else will reply immeditately.

We'd still need to change device-steps.sh if this approach unless we want the zedagent etc start processes to sit around in the background and wait for the done forever; we need to get the start to complete and not consume memory.

File: pkg/pillar/scripts/device-steps.sh
Comment (deitch) [reply pkg/pillar/scripts/device-steps.sh:10]: +2

File: pkg/pillar/scripts/device-steps.sh
Comment (eriknordmark) [reply pkg/pillar/scripts/device-steps.sh:65]: We can have invoke a 10-40 Mbyte go binary to wait for something on pubsub. A bit simpler than the waitforaddr.go we have to wait for nim to get IP addresses. We can make such a waitforagentdone smaller since it doesn't need http etc, If we do we can replace the wait for the agent starting in the script (which looks for the .touch file) as well.

File: pkg/pillar/zedbox/zedbox.go
Comment (eriknordmark) [inline pkg/pillar/zedbox/zedbox.go:95]: I assume we don't want this periodic dump in the product.
Can we make SIGUSR2 log the profile so we can get it in Kibana? Or do we only want this to be dumped locally on demand in the product?

File: pkg/pillar/zedbox/zedbox.go
Comment (eriknordmark) [inline pkg/pillar/zedbox/zedbox.go:113]: Instead of partial duplication of pubsub, would it make sense to define a service+args golang struct and use that with pubsub to tell zedbox what to spawn? That assumes that the code which starts the services is a golang app (perhaps a tiny one only importing pubsub).

File: pkg/pillar/cmd/tpmmgr/tpmmgr.go
Comment (adarsh-zededa) [reply pkg/pillar/cmd/tpmmgr/tpmmgr.go]: Done.

File: pkg/pillar/scripts/device-steps.sh
Comment (adarsh-zededa) [reply pkg/pillar/scripts/device-steps.sh]: Done.

File: pkg/pillar/Makefile
Comment (adarsh-zededa) [reply pkg/pillar/Makefile]: Reverted.

File: pkg/pillar/zedbox/zedbox.go
Comment (eriknordmark) [inline pkg/pillar/zedbox/zedbox.go:106]: Why not put the zedbox Run code in a Run() function so you don't need to look for debug etc twice?
Basically just add "zedbox": Run to the entryPoints definition.

File: pkg/pillar/zedbox/zedbox.go
Comment (eriknordmark) [inline pkg/pillar/zedbox/zedbox.go:160]: Instead of invoking it directly, why not invoke a wrapper which writes agentName.done?
Something like
func runAndDone(agentName string, ps *pubsub.Pubsub) {
     entrypoint(ps)
     log.Infof("service %s done", agentName)
     writeDoneFile(agentnane)
}

That way you avoid code in any of the agents for the done part, and you don't need to define specific filenames in the types package.

File: pkg/pillar/scripts/device-steps.sh
Comment (eriknordmark) [reply pkg/pillar/scripts/device-steps.sh:66]: As I said above, I think this should be called wait_for_done $agentname $time

Comment (adarsh-zededa) [reply pkg/pillar/scripts/device-steps.sh:81]: Added `wait_for_done()`.
Had to retain `wait_for_file()` to check on `/config/uuid`.

File: pkg/pillar/zedbox/zedbox.go
Comment (adarsh-zededa) [reply pkg/pillar/zedbox/zedbox.go]: Changed.

File: pkg/pillar/zedbox/zedbox.go
Comment (adarsh-zededa) [reply pkg/pillar/zedbox/zedbox.go]: Removed this considering memory usage.

File: pkg/grub/rootfs.cfg
Comment (adarsh-zededa) [reply pkg/grub/rootfs.cfg]: @rvs, I tested with 300M softLimit and 400M hardLimit for EVE services.
The device was stable (with 5 dataahead containers and 1 redis container).
But had watchdog issues while updating EVE-image. 
I'm finetuning the memory further.

File: pkg/pillar/scripts/device-steps.sh
Comment (eriknordmark) [reply pkg/pillar/scripts/device-steps.sh:81]: Can't you instead wait for client.done?

File: pkg/pillar/scripts/device-steps.sh
Comment (eriknordmark) [inline pkg/pillar/scripts/device-steps.sh:222]: This should be infinite as well to match current behavior

File: pkg/pillar/scripts/device-steps.sh
Comment (eriknordmark) [inline pkg/pillar/scripts/device-steps.sh:171]: Infinite wait

File: pkg/pillar/scripts/device-steps.sh
Comment (eriknordmark) [inline pkg/pillar/scripts/device-steps.sh:418]: Infinite wait

File: pkg/pillar/scripts/device-steps.sh
Comment (eriknordmark) [inline pkg/pillar/scripts/device-steps.sh:425]: Infinite

File: pkg/pillar/scripts/device-steps.sh
Comment (eriknordmark) [inline pkg/pillar/scripts/device-steps.sh:449]: Why can't you wait for client.done here?

File: pkg/pillar/zedbox/zedbox.go
Comment (eriknordmark) [inline pkg/pillar/zedbox/zedbox.go:269]: Question; if the agent aka zedbox hits a log.Fatal or panic I assume the serviceEntrypoint aka Run() will not return. Is that correct?

File: pkg/pillar/zedbox/zedbox.go
Comment (eriknordmark) [inline pkg/pillar/zedbox/zedbox.go:274]: /var/run/%s.done, right?

**Review by eriknordmark — CHANGES_REQUESTED** [review]
You shouldn't need the definitions for the done files in types package, nor the code to write that file in the agents. Only needed code is in zedbox itself plus the script looking for the done files.

So some code can be deleted. Note that the zedbox code doesn't write the done file to the correct path.

I don't know about the 600 seconds vs. infinite wait. If any service (client or tpmmgr) does a panic or log.Fatal we want to bail. But since that will cause a zedbox crash the watchdog will catch it.
So AFAICT all the wait_for_done can be forever. Does that make sense?

File: pkg/pillar/scripts/device-steps.sh
Comment (rvs) [reply pkg/pillar/scripts/device-steps.sh:81]: Hey @eriknordmark -- I think in general I'd prefer wrapping all this into a pubsub -- as you're saying we can make sure it is a pretty small standalone CLI that would allow anything to walk on the pubsub.

That said, I'd rather see this PR merged as-is and we can quickly follow up with pubsub refactoring.

Does this make sense to you @eriknordmark and @adarsh-zededa ?

**Comment by eriknordmark** [comment]
> LGTM -- lets get this in quickly and follow up with a few bits of targeted refactoring.

Yeah I'll touch base with @adarsh-zededa tomorrow morning and we can see if there is something more underway for this PR or we can pull it tomorrow. We can fine tune as we go along.

File: pkg/pillar/zedbox/zedbox.go
Comment (eriknordmark) [reply pkg/pillar/zedbox/zedbox.go]: @rvs Once this is in master we can do a config file or something which can specify how many binaries to run and the mapping from microservices to zedbox processes. Shouldn't be too hard given the infrastructure in this PR

**Review by eriknordmark — COMMENTED** [review]
Does $BINDIR/waitforaddr actually wait, or do we need a wait_for_done waitforaddr after it in device-steps.sh?

Note that if we don't wait it only appears during onboarding that we will not get ntp synchronized before we create a device certificate with an incorrect timestamp in it. I think this is the only mustfix before we pull.

File: pkg/pillar/zedbox/zedbox.go
Comment (adarsh-zededa) [reply pkg/pillar/zedbox/zedbox.go:269]: Yeah.

File: pkg/pillar/zedbox/zedbox.go
Comment (adarsh-zededa) [reply pkg/pillar/zedbox/zedbox.go]: Corrected. Thanks!

File: pkg/pillar/scripts/device-steps.sh
Comment (adarsh-zededa) [reply pkg/pillar/scripts/device-steps.sh]: I assumed that was one of the forever running agent.
Changed.

File: pkg/pillar/scripts/device-steps.sh
Comment (adarsh-zededa) [reply pkg/pillar/scripts/device-steps.sh]: Done.

File: pkg/pillar/scripts/device-steps.sh
Comment (adarsh-zededa) [reply pkg/pillar/scripts/device-steps.sh]: Done.

File: pkg/pillar/scripts/device-steps.sh
Comment (adarsh-zededa) [reply pkg/pillar/scripts/device-steps.sh]: Done.

File: pkg/pillar/scripts/device-steps.sh
Comment (adarsh-zededa) [reply pkg/pillar/scripts/device-steps.sh]: Done.

**Review by eriknordmark — APPROVED** [review]

**Pull request merged**
